File: Components/DevExpressVCL/Demos/ExpressBars/CBuilder/InPlaceEditorsDemo/InPlaceEditorsDemoFrameManager.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "InPlaceEditorsDemoFrameManager.h"
#include <Math.hpp>
#include "cxGraphics.hpp"
#include "cxGeometry.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxGraphics"
#pragma link "cxGeometry"
#pragma resource "*.dfm"
TEditorDemoBaseFrame *EditorDemoBaseFrame;
TEditorDemoFrameManager* FInstance;
//---------------------------------------------------------------------------

void __fastcall TEditorDemoBaseFrame::DrawText(AnsiString AStrings[], TPaintBox* APaintBox, int AStringsCount)
{
  String S = "";
  TSize ASize = Size(0, 0);
  for (int I = 0; I <= (AStringsCount - 1); I++)
  {
    ASize.cx = Max(ASize.cx, APaintBox->Canvas->TextWidth(AStrings[I]));
    S = S + AStrings[I] + "\n";
  };
  ASize.cy = APaintBox->Canvas->TextHeight("Wg") * AStringsCount;
  TRect ARect = cxRectCenter(APaintBox->ClientRect, ASize);
  cxDrawText(THandle(APaintBox->Canvas->Handle), S, ARect, DT_WORDBREAK, -1);
}
//---------------------------------------------------------------------------
__fastcall TEditorDemoBaseFrame::TEditorDemoBaseFrame(TComponent* Owner)
	: TForm(Owner)
{

}
//---------------------------------------------------------------------------
bool __fastcall TEditorDemoBaseFrame::GetDescriptionVisible()
{
  return(lblFrameDescription->Visible);
}
//---------------------------------------------------------------------------
void __fastcall TEditorDemoBaseFrame::SetDescription(AnsiString ADescription)
{
  lblFrameDescription->Caption = ADescription;
};
//---------------------------------------------------------------------------
void __fastcall TEditorDemoBaseFrame::SetDescriptionVisible(bool AValue)
{
  lblFrameDescription->Visible = AValue;
};
//---------------------------------------------------------------------------
__fastcall TEditorDemoFrameManager::TEditorDemoFrameManager()
{
  FFrameList = new TList();
};
//---------------------------------------------------------------------------
__fastcall TEditorDemoFrameManager::~TEditorDemoFrameManager()
{
  delete FFrameList;
};
//---------------------------------------------------------------------------
TEditorDemoBaseFrame* __fastcall TEditorDemoFrameManager::GetFrame(int AIndex)
{
  TEditorDemoBaseFrame* Result;
  for(int I = 0; I <= (FramesCount - 1); I++)
  {
    Result = (TEditorDemoBaseFrame*)(FFrameList->Items[I]);
    if (Result->Tag == AIndex)
    {
      return(Result);
    };
  };
  return(0);
};
//---------------------------------------------------------------------------
void __fastcall TEditorDemoFrameManager::UpdateFrameColors(TColor AColor)
{
  for(int I = 0; I <= (FFrameList->Count - 1); I++)
    Frames[I]->Color = AColor;
};
//---------------------------------------------------------------------------
void __fastcall TEditorDemoFrameManager::AddFrame(TForm* AFrame, int AFrameId)
{
  FFrameList->Add(AFrame);
  AFrame->Tag = AFrameId;
};
//---------------------------------------------------------------------------
void __fastcall TEditorDemoFrameManager::SetDescriptionsVisible(bool AValue)
{
  for (int I = 0; I <= (FFrameList->Count - 1); I++)
    Frames[I]->SetDescriptionVisible(AValue);
};
//---------------------------------------------------------------------------
int __fastcall TEditorDemoFrameManager::GetFramesCount()
{
  return(FFrameList->Count);
};
//---------------------------------------------------------------------------
TEditorDemoFrameManager* EditorDemoFrameManager()
{
  if (!FInstance)
    FInstance = new TEditorDemoFrameManager();
  return FInstance;
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressBars/Delphi/InPlaceEditorsDemo/InPlaceEditorsDemoFrameManager.pas
unit InPlaceEditorsDemoFrameManager;

interface

uses
  Windows, Classes, Controls, SysUtils, Graphics, Forms, StdCtrls, ExtCtrls;

type
  TEditorDemoBaseFrame = class(TForm)
    lblFrameDescription: TLabel;
  private
    function GetDescriptionVisible: Boolean;
    procedure SetDescriptionVisible(AValue: Boolean);
  public
    procedure SetDescription(ADescription: string);

    property DescriptionVisible: Boolean read GetDescriptionVisible write SetDescriptionVisible;
  end;

  TEditorDemoFrameManager = class
  private
    FFrameList: TList;
    function GetFrame(AIndex: Integer): TEditorDemoBaseFrame;
    function GetFramesCount: Integer;
  public
    constructor Create; virtual;
    destructor Destroy; override;
    procedure AddFrame(AFrame: TForm; AFrameId: Integer);
    procedure UpdateFrameColors(AColor: TColor);
    procedure SetDescriptionsVisible(AValue: Boolean);

    property FramesCount: Integer read GetFramesCount;
    property Frames[AIndex: Integer]: TEditorDemoBaseFrame read GetFrame; default;
  end;

procedure DrawText(AStrings: array of string; APaintBox: TPaintBox);

var
  AFrameManager: TEditorDemoFrameManager;

implementation

{$R *.dfm}

uses
  Types, Math, cxGraphics, cxGeometry;

procedure DrawText(AStrings: array of string; APaintBox: TPaintBox);
var
  I: Integer;
  S: string;
  ARect: TRect;
  ASize: TSize;
begin
  S := '';
  ASize.cx := 0;
  for I := Low(AStrings) to High(AStrings) do
  begin
    ASize.cx := Max(ASize.cx, APaintBox.Canvas.TextWidth(AStrings[I]));
    S := S + AStrings[I] + #13#10;
  end;

  ASize.cy := APaintBox.Canvas.TextHeight('Wg') * Length(AStrings);

  ARect := cxRectCenter(APaintBox.ClientRect, ASize);
  cxDrawText(APaintBox.Canvas.Handle, S, ARect, DT_WORDBREAK);
end;

{ TEditorDemoBaseFrame }

procedure TEditorDemoBaseFrame.SetDescription(ADescription: string);
begin
  lblFrameDescription.Caption := ADescription;
end;

function TEditorDemoBaseFrame.GetDescriptionVisible: Boolean;
begin
  Result := lblFrameDescription.Visible;
end;

procedure TEditorDemoBaseFrame.SetDescriptionVisible(AValue: Boolean);
begin
  lblFrameDescription.Visible := AValue;
end;

{ TEditorDemoFrameManager }

constructor TEditorDemoFrameManager.Create;
begin
  FFrameList := TList.Create;
end;

destructor TEditorDemoFrameManager.Destroy;
begin
  FreeAndNil(FFrameList);
  inherited;
end;

procedure TEditorDemoFrameManager.AddFrame(AFrame: TForm; AFrameId: Integer);
begin
  FFrameList.Add(AFrame);
  AFrame.Tag := AFrameId;
end;

function TEditorDemoFrameManager.GetFrame(
  AIndex: Integer): TEditorDemoBaseFrame;
var
  I: Integer;
begin
  Result := nil;
  for I := 0 to FramesCount - 1 do
    if TEditorDemoBaseFrame(FFrameList[I]).Tag = AIndex then
    begin
      Result := TEditorDemoBaseFrame(FFrameList[I]);
      Break;
    end;
end;

function TEditorDemoFrameManager.GetFramesCount: Integer;
begin
  Result := FFrameList.Count;
end;

procedure TEditorDemoFrameManager.UpdateFrameColors(AColor: TColor);
var
  I: Integer;
begin
  for I := 0 to FFrameList.Count - 1 do
    Frames[I].Color := AColor;
end;

procedure TEditorDemoFrameManager.SetDescriptionsVisible(AValue: Boolean);
var
  I: Integer;
begin
  for I := 0 to FramesCount - 1 do
    Frames[I].SetDescriptionVisible(AValue);
end;

initialization
  AFrameManager := TEditorDemoFrameManager.Create;

finalization
  AFrameManager.Free;
end.

